pkg/controllers/provisioning/scheduling/topologynodefilter.go

```diff
@@ -27,102 +27,60 @@ import (
 // included for topology counting purposes. This is only used with topology spread constraints as affinities/anti-affinities
 // always count across all nodes. A nil or zero-value TopologyNodeFilter behaves well and the filter returns true for
 // all nodes.
-type TopologyNodeFilter struct {
-	nodeSelector      map[string]string
-	nodeSelectorTerms []v1.NodeSelectorTerm
-}
+type TopologyNodeFilter []v1alpha5.Requirements
 
-func NewTopologyNodeFilter(p *v1.Pod) *TopologyNodeFilter {
-	selector := &TopologyNodeFilter{
-		nodeSelector: p.Spec.NodeSelector,
-	}
-	if p.Spec.Affinity != nil && p.Spec.Affinity.NodeAffinity != nil && p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
-		selector.nodeSelectorTerms = p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+func MakeTopologyNodeFilter(p *v1.Pod) TopologyNodeFilter {
+	nodeSelectorRequirements := v1alpha5.NewLabelRequirements(p.Spec.NodeSelector)
+	// if we only have a label selector, that's the only requirement that must match
+	if p.Spec.Affinity == nil || p.Spec.Affinity.NodeAffinity == nil || p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return TopologyNodeFilter{nodeSelectorRequirements}
 	}
-	return selector
-}
 
-// Matches returns true if the TopologyNodeFilter doesn't prohibit node from the participating in the topology
-func (t *TopologyNodeFilter) Matches(node *v1.Node) bool {
-	if t == nil {
-		return true
-	}
-	// if a node selector term is provided, it must match the node
-	for k, v := range t.nodeSelector {
-		if node.Labels[k] != v {
-			return false
-		}
-	}
-	if len(t.nodeSelectorTerms) == 0 {
-		return true
+	// otherwise, we need to match the combination of label selector and any term of the required node affinities since
+	// those terms are OR'd together
+	var filter TopologyNodeFilter
+	for _, term := range p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		requirements := nodeSelectorRequirements.Add(v1alpha5.NewRequirements(term.MatchExpressions...).Requirements...)
+		filter = append(filter, requirements)
 	}
 
-	for _, term := range t.nodeSelectorTerms {
-		requirement := v1alpha5.NewRequirements(term.MatchExpressions...)
-		if t.matchesRequirements(requirement, node) {
-			return true
-		}
-	}
+	return filter
+}
 
-	// at this point we have node selector terms, but didn't match all of the requirements for any individual term
-	return false
+// Matches returns true if the TopologyNodeFilter doesn't prohibit node from the participating in the topology
+func (t TopologyNodeFilter) Matches(node *v1.Node) bool {
```

Comment:
This is used in one place. Thoughts on removing this and just using the Compatible interface in that case? 
```
nodeFilter.Compatible(v1alpha5.NewLabelRequirements(node.Labels))
```

Reply:
The concept of a filter being "Compatible" with a node is less intuitive to me than the concept of a filter "Matching" a node.